beeline.go

```diff
@@ -97,65 +102,146 @@ func Init(config Config) {
+	AddFieldToSpan(ctx, key, val)
 }
 
-// ContextWithEvent returns a new context created from the passed context with a
-// Honeycomb event added to it. In most cases, the code adding the event to the
-// context should also be responsible for sending that event on to Honeycomb
-// when it's finished.
-func ContextWithEvent(ctx context.Context, ev *libhoney.Event) context.Context {
-	return context.WithValue(ctx, honeyEventContextKey, ev)
+// AddFieldToSpan allows you to add a single field to an event anywhere
+// downstream of an instrumented request. After adding the appropriate
+// middleware or wrapping a Handler, feel free to call AddFieldToSpan freely
+// within your code. Pass it the context from the request (`r.Context()`) and
+// the key and value you wish to add.This function is good for span-level data,
+// eg timers or the arguments to a specific function call, etc. Fields added
+// here are prefixed with `app.`
+func AddFieldToSpan(ctx context.Context, key string, val interface{}) {
+	namespacedKey := fmt.Sprintf("app.%s", key)
+	span := internal.CurrentSpan(ctx)
+	if span != nil {
+		span.AddField(namespacedKey, val)
+	}
 }
 
-// ContextEvent retrieves the Honeycomb event from a context. You can add fields
-// to the event or override settings (eg sample rate) but should not Send() the
-// event; the wrapper that inserted the event into the Context is responsible
-// for sending it to Hnoeycomb
-func ContextEvent(ctx context.Context) *libhoney.Event {
-	if ctx != nil {
-		if evt, ok := ctx.Value(honeyEventContextKey).(*libhoney.Event); ok {
-			return evt
-		}
+// AddRollupFieldToSpan allows you to add a numeric field to the current span,
+// and, when called on multiple spans within a trace, the sum of the field will
+// be added to the root span. Use this when doing an action many times or on
+// many spans and you want the sum of all those actions to be represented on the
+// root span. Fields added here are prefixed with `app.` and the rolled up
+// fields on the root span are prefixed with `totals.app.`
+func AddRollupFieldToSpan(ctx context.Context, key string, val float64) {
+	namespacedKey := fmt.Sprintf("app.%s", key)
+	span := internal.CurrentSpan(ctx)
+	if span != nil {
+		span.AddRollupField(namespacedKey, val)
 	}
-	return nil
 }
 
-// contextBuilder isn't used yet but matches ContextEvent. When it's useful,
-// export it, but until then it's just confusing.
-func contextBuilder(ctx context.Context) *libhoney.Builder {
-	if bldr, ok := ctx.Value(honeyBuilderContextKey).(*libhoney.Builder); ok {
-		return bldr
+// AddFieldToTrace adds the field to both the currently active span and all
+// other spans involved in this trace that occur within this process.
+// Additionally, these fields are packaged up and passed along to downstream
+// processes if they are also using a beeline. This function is good for adding
+// context that is better scoped to the request than this specific unit of work,
```

Comment:
This is a bit confusing - haven't we previously used "processing a web request" as an example of what a unit of work _is_? More generally - what corresponds to a unit of work in the Beeline / tracing world? Is it a trace, a span, the root span specifically, or something else?